Lib/functools.py

```diff
@@ -990,15 +991,17 @@ def __get__(self, instance, owner=None):
                 f"instance to cache {self.attrname!r} property."
             )
             raise TypeError(msg) from None
-        val = self.func(instance)
-        try:
-            cache[self.attrname] = val
-        except TypeError:
-            msg = (
-                f"The '__dict__' attribute on {type(instance).__name__!r} instance "
-                f"does not support item assignment for caching {self.attrname!r} property."
-            )
-            raise TypeError(msg) from None
+        val = cache.get(self.attrname, _NOT_FOUND)
+        if val is _NOT_FOUND:
+            val = self.func(instance)
+            try:
+                cache[self.attrname] = val
+            except TypeError:
+                msg = (
+                    f"The '__dict__' attribute on {type(instance).__name__!r} instance "
+                    f"does not support item assignment for caching {self.attrname!r} property."
+                )
+                raise TypeError(msg) from None
```

Comment:
How about not adding _NOT_FOUND object?

```suggestion
        if self.attrname in cache:
            return cache[self.attrname]
        else:
            val = self.func(instance)
            try:
                cache[self.attrname] = val
            except TypeError:
                msg = (
                    f"The '__dict__' attribute on {type(instance).__name__!r} instance "
                    f"does not support item assignment for caching {self.attrname!r} property."
                )
                raise TypeError(msg) from None
        return val
```

Reply:
Is there a problem with adding the `_NOT_FOUND` object?

The suggested version is slightly less efficient because (in the "found" case) we have to do two dictionary lookups instead of only one. It's not a big deal, but I don't see any strong reason to do this.

Reply:
> Is there a problem with adding the _NOT_FOUND object?

Yeah, it's not a strong opinion, I just avoid adding new object as possible but if you prefer this approach I am fine with :)